### PR TITLE
kubeadm: prompt for confirmation when resetting a master

### LINKED ISSUE #673: Require user confirmation when resetting a master node
### FEATURE REQUEST

Using `kubeadm` 1.9, I accidentally issued a `kubeadm reset` on my master node instead of a worker node (due to a dropped SSH connection). This deleted the cluster credentials and despite the `etcd` manifest using a `hostDir` for data, it seems that the re-created cluster did not successfully get repopulated with the previous etcd data.

I think it would be really nice if there was a strong confirmation requirement when `kubeadm reset` is issued on a master node. Something like:
```
$ kubeadm reset
[Warning] Are you sure you want to delete this master node? Type the word 'confirm' below to continue:
```

Luckily it is/was just a test cluster, but still caused a bit of (avoidable) pain...

### PR DESCRIPTION
**Pull request by alexbrand**
Signed-off-by: Alexander Brand <alexbrand09@gmail.com>

**What this PR does / why we need it**:
This PR implements a confirmation prompt when running `kubeadm reset` on a master node. This is to prevent users from mistakenly resetting a master node.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/673

**Special notes for your reviewer**:
I am somewhat torn on the approach on how to detect that kubeadm is running on a master node. I went with checking for the apiserver manfiest file on the local filesystem, as it seems like a simpler approach when compared to getting a k8s client, getting a list of nodes, finding the current node, and checking if it has the master taint. I am happy to rework if the latter is more desirable.

Sample runs:
```
# ./kubeadm reset
[Warning] Are you sure you want to reset this master node? Type the word "confirm" to continue: no
Aborted reset operation on master node

# ./kubeadm reset
[Warning] Are you sure you want to reset this master node? Type the word "confirm" to continue: confirm
[preflight] Running pre-flight checks.
[reset] Stopping the kubelet service.
[reset] WARNING: The kubelet service could not be stopped by kubeadm: [exit status 1]
[reset] WARNING: Please ensure kubelet is stopped manually.
[reset] Unmounting mounted directories in "/var/lib/kubelet"
........

# ./kubeadm reset
[Warning] Are you sure you want to reset this master node? Type the word "confirm" to continue:
Aborted reset operation on master node
# ./kubeadm reset --confirm
[preflight] Running pre-flight checks.
[reset] Stopping the kubelet service.
[reset] WARNING: The kubelet service could not be stopped by kubeadm: [exit status 1]
[reset] WARNING: Please ensure kubelet is stopped manually.
[reset] Unmounting mounted directories in "/var/lib/kubelet"
........
```

**Release note**:

```release-note
kubeadm: prompt the user for confirmation when resetting a master node
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/reset.go [hunk 1]
```diff
@@ -82,6 +87,11 @@ func NewCmdReset(out io.Writer) *cobra.Command {
 		"The path to the CRI socket to use with crictl when cleaning up containers.",
 	)
 
+	cmd.PersistentFlags().BoolVar(
+		&userConfirmedReset, "confirm", false,
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/reset.go]: What about --force or --silent instead?

**Review by fabriziopandini — COMMENTED** [review]
Thanks for this PR!

I like the idea; I'm just wondering if it could be useful to prompt always (no matter of the type of node)? Actually also resetting a node by mistake could be annoying...
WDYT?

**Comment by alexbrand** [comment]
Thanks for the review @fabriziopandini.

Interestingly, I considered prompting for any node type, but I convinced myself that the user typing `kubeadm reset` is a good enough indication that the reset is desired. Happy to discuss it more though, and get more feedback from others.

File: cmd/kubeadm/app/cmd/reset.go (hunk 1 above)
Comment (alexbrand) [reply cmd/kubeadm/app/cmd/reset.go]: Heh, I had force initially. I don't have a strong opinion here, so happy to make the change if we think that's better.

**Comment by fabriziopandini** [comment]
Ok 
Let's get more feedbacks from the others.
In the meantime one last question from my side: I don't think the current approach for detecting master node could work in case of self-hosted control plane. Have you considered this scenario?

**Comment by alexbrand** [comment]
Ah, yes. That is a very good point. I think we'll have to connect to the API server and get the node resource to figure out if it's a master node (via annotation). Thanks @fabriziopandini

**Comment by fabriziopandini** [comment]
@timothysc @errordeveloper  WDYT if `kubeadm reset` always prompts for confirmation - no matter of the type of node - ? This will provide a consistent UX and simplify this PR ...
Nb automation tools can use --force

**Comment by binarybana** [comment]
I can see the argument for confirming on all nodes, but I think in that case, the confirmation for head nodes should be made different somehow, to avoid reflexive confirmation when you think you're on a worker node but are accidentally on a master node (this happened to me recently and was the original motivation that prompted this feature request).

I personally think resetting on worker nodes doesn't need confirmation (they're cattle after all). But I do administer my cluster with `xpanes` (tmux panes for each worker), so ya...

**Comment by alexbrand** [comment]
@fabriziopandini Indeed - prompting on all nodes will significantly reduce the complexity of this change. I think I like the idea, but wondering what others think

File: cmd/kubeadm/app/cmd/reset.go [hunk 2]
```diff
@@ -262,3 +276,32 @@ func resetConfigDir(configPathDir, pkiPathDir string) {
...
+	_, err := os.Stat(apiServerManifest)
+	if os.IsNotExist(err) {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("could not determine if this is a master node: %v", err)
+	}
+
+	// This is a master node - prompt the user for confirmation
+	fmt.Printf("[Warning] Are you sure you want to reset this master node? Type the word %q to continue: ", confirmationWord)
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/reset.go]: I'd prefer just a [y/n] response here.  Also we might want to check is kubectl has any blocking interactions and copy the pattern.  

If they do I would be surprised if there wasn't already a utility for it. 

/cc @pwittrock

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/reset.go (hunk 2 above)
Comment (alexbrand) [reply cmd/kubeadm/app/cmd/reset.go]: Sounds good. I'll look to see if there are blocking interactions in kubectl that we can be consistent with.

**Comment by alexbrand** [comment]
@timothysc Thanks for the review. There's an open question raised by @fabriziopandini on control plane node detection when running a self-hosted cluster, given that the apiserver manfiest file will not be available. 

IIRC, previous discussion suggested accessing the API server and using annotations on the node resource to determine if the node is a control plane node, but this is problematic because kubeadm does not know the node name of the machine where it is running (without introducing a new requirement on `kubeadm reset` to read the kubeadm config file).

**Comment by timothysc** [comment]
@alexbrand 

Confirming reset on * is the easiest solution and is consistent.  I don't think you should special case the master nodes, we could but the extra logic is not necessary to have a consistent UX.

**Comment by alexbrand** [comment]
Aiming to have this complete by tomorrow.

**Comment by alexbrand** [comment]
@timothysc Updated the prompt to `[y/N]` style, and also prompting on all nodes now. I looked around in kubectl for any blocking interactions but could not find any.

File: cmd/kubeadm/app/cmd/reset.go [hunk 3]
```diff
@@ -93,9 +101,21 @@ type Reset struct {
...
+func NewReset(in io.Reader, ignorePreflightErrors sets.String, forceReset bool, certsDir, criSocketPath string) (*Reset, error) {
+	if !forceReset {
+		fmt.Println("[reset] WARNING: changes made to this host by 'kubeadm init' or 'kubeadm join' will be reverted.")
+		fmt.Print("[reset] are you sure you want to proceed? [y/N]: ")
+		s := bufio.NewScanner(in)
+		s.Scan()
+		if err := s.Err(); err != nil {
+			return nil, err
+		}
+		if s.Text() != "y" {
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/reset.go]: Don't you want to tolower and ignore spaces and empty returns?

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/reset.go (hunk 3 above)
Comment (alexbrand) [reply cmd/kubeadm/app/cmd/reset.go]: Thanks. I'll add tolower so that it is less finicky. 

Spaces and empty returns will abort.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**